rust/src/buffer.rs

```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
         self.len
     }
 
+    /// Get a pointer to the data contained by the buffer
     pub fn data(&self) -> *const T {
         self.data
     }
 
     pub fn slice(&self, start: usize, end: usize) -> &[T] {
         assert!(start <= end);
-        assert!(start <= self.len as usize);
+        assert!(start < self.len as usize);
         assert!(end <= self.len as usize);
         unsafe { slice::from_raw_parts(self.data.offset(start as isize), (end - start) as usize) }
     }
 
     /// Get a reference to the value at the specified offset
     pub fn get(&self, i: usize) -> &T {
+        assert!(i < self.len as usize);
```

Comment:
panic message would be useful - `assert!(..., "Out-of-bounds access")` (same goes for elsewhere)